activesupport/lib/active_support/cache/key_master.rb

```diff
@@ -0,0 +1,165 @@
+    #   k2 = KeyMaster.new(k1)
+    #   k1.cache_key == k2.cache_key
+    #   # => true
+    #
+    # In addition to passing a key into the initialize method
+    # key fragments can be added onto an existing key through the
+    # `<<` method.
+    #
+    # Example:
+    #
+    #   key = KeyMaster.new
+    #   key << "who-ya-gonna-call"
+    #   key << "ghost-busters"
+    #   key.cache_key
+    #   # => "who-ya-gonna-call/ghost-busters"
+    #
+    # Multiple key entries generate the same key regardless of how
+    # they are added. I.e. if they're added via multiple arrays
+    # or directly via the `<<` operator.
+    #
+    # Example:
+    #
+    #   k1 = KeyMaster.new("slimer")
+    #   k1 << "stay-puft"
+    #   k2 = KeyMaster.new(["slimer", "stay-puft"])
+    #   k1.cache_key == k2.cache_key
+    #   # => true
+    #   k3 = KeyMaster.new(["slimer", ["stay-puft"]])
+    #   k2.cache_key == k3.cache_key
+    #   # => true
+    #
+    # It is faster to directly add objects to the
+    # KeyMaster object via `<<` than to generate
+    # an intermediate array of objects
+    #
+    # The class supports strings, arrays, hashes,
+    # objects with a `cache_key` method and anything
+    # that can be implicitly cast to a string:
+    #
+    # Examples:
+    #
+    #   KeyMaster.new("there").cache_key
+    #   # => "there"
+    #
+    #   KeyMaster.new(["is", "no"]).cache_key
+    #   # => "is/no"
+    #
+    #   KeyMaster.new({ dana: 1, only: 2 }).cache_key
+    #   # => "dana=1/only=2"
+    #
+    #   klass = Class.new { def cache_key; "zuul"; end }
+    #   KeyMaster.new(klass.new).cache_key
+    #   # => "zuul"
+    #
+    # When generating a a cache key, `nil` will be filtered
+    # out, so `["ghost", nil "busters"]` will generate the
+    # same output as ["ghost", "busters"]. If presence of a
+    # `nil` is important, then convert it to a string.
+    #
+    # In addition to generating a `cache_key`
+    # a `cache_version` is also generated if the object
+    # passed in responds to `cache_version` or if it contains
+    # an element that responds to `cache_version`.
+    #
+    # The purpose of keeping this value seperate from
+    # `cache_key` is to support the use of "cache versioning"
+    # where the same key is recycled but the version of the object
+    # being stored is kept directly in the cache.
+    #
+    # Example:
+    #
+    #   klass = Class.new { def cache_version; "patty"; end }
+    #   KeyMaster.new(klass.new).cache_version
+    #   # => "patty"
+    #
+    #   klass = Class.new { def cache_version; "egon"; end }
+    #   KeyMaster.new([klass.new]).cache_version
+    #   # => "egon"
+    #
+    #   klass = Class.new { def cache_version; "venkman"; end }
+    #   KeyMaster.new({foo: klass.new}).cache_version
+    #   # => "venkman"
+    class KeyMaster # :nodoc:
+      attr_reader :cache_key, :cache_version
+
+      def self.to_key(key)
+        return key if key.is_a?(KeyMaster)
+
+        self.new(key)
+      end
```

Comment:
Would compare before/after usage with `.for(key)` or simply `.new(key)`. The `to_*` methods indicate that the receiver is being coerced, not the argument.

No need for a guard here, since it's expected that a key may be an instance of this class. A conditional expression is appropriate.